File: app/src/main/java/com/nguyendevs/stkh/Settings/SettingsActivity.java
package com.nguyendevs.stkh.Settings;

import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.widget.*;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import com.nguyendevs.stkh.Database.DatabaseHelper;
import com.nguyendevs.stkh.MainActivity;
import com.nguyendevs.stkh.R;

public class SettingsActivity extends AppCompatActivity {
    private String selectedLanguage = "Vietnamese";
    private String selectedModel = "Medium";
    private int speechSpeed = 50;
    private int speechPitch = 50;
    private DatabaseHelper dbHelper;
    private Spinner languageSpinner, modelSpinner;
    private TextView modelLabel;
    private SharedPreferences prefs;
    private EditText serverIP;
    private SeekBar speechSpeedSeekBar, speechPitchSeekBar;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.settinglayout);
        serverIP = findViewById(R.id.serverip);
        dbHelper = new DatabaseHelper(this);
        prefs = getSharedPreferences("AppPrefs", MODE_PRIVATE);
        selectedLanguage = prefs.getString("selectedLanguage", "Vietnamese");
        selectedModel = prefs.getString("selectedModel", "Medium");
        speechSpeed = prefs.getInt("speechSpeed", 50);
        speechPitch = prefs.getInt("speechPitch", 50);
        serverIP.setText(prefs.getString("serverIP", "14.235.42.182"));
        serverIP.addTextChangedListener(new TextWatcher() {
            public void afterTextChanged(Editable s) {
                prefs.edit().putString("serverIP", s.toString().trim()).apply();
            }
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {}
            public void onTextChanged(CharSequence s, int start, int before, int count) {}
        });
        languageSpinner = findViewById(R.id.language);
        updateLanguageSpinner();
        modelSpinner = findViewById(R.id.model);
        modelLabel = findViewById(R.id.modeler);
        String[] models = {"Small", "Medium", "Large"};
        ArrayAdapter<String> modelAdapter = new ArrayAdapter<>(this, android.R.layout.simple_spinner_dropdown_item, models);
        modelSpinner.setAdapter(modelAdapter);
        modelSpinner.setSelection(java.util.Arrays.asList(models).indexOf(selectedModel));
        modelSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                selectedModel = models[position];
                prefs.edit().putString("selectedModel", selectedModel).apply();
            }
            public void onNothingSelected(AdapterView<?> parent) {}
        });
        modelSpinner.setVisibility(View.VISIBLE);
        modelLabel.setVisibility(View.VISIBLE);
        speechSpeedSeekBar = findViewById(R.id.speech_speed);
        speechSpeedSeekBar.setProgress(speechSpeed);
        speechSpeedSeekBar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                speechSpeed = progress;
                prefs.edit().putInt("speechSpeed", speechSpeed).apply();
            }
            public void onStartTrackingTouch(SeekBar seekBar) {}
            public void onStopTrackingTouch(SeekBar seekBar) {}
        });
        speechPitchSeekBar = findViewById(R.id.speech_pitch);
        speechPitchSeekBar.setProgress(speechPitch);
        speechPitchSeekBar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                speechPitch = progress;
                prefs.edit().putInt("speechPitch", speechPitch).apply();
            }
            public void onStartTrackingTouch(SeekBar seekBar) {}
            public void onStopTrackingTouch(SeekBar seekBar) {}
        });

        findViewById(R.id.clear_history_button).setOnClickListener(v -> showClearHistoryDialog());
        findViewById(R.id.reset_button).setOnClickListener(v -> showResetDialog());
        findViewById(R.id.back).setOnClickListener(v -> returnToMainActivity());
        findViewById(R.id.about).setOnClickListener(v -> showAboutDialog());
    }
    private void showAboutDialog() {
        new AlertDialog.Builder(this)
                .setTitle("Về chúng tôi")
                .setMessage("Ứng dụng được phát triển bởi NguyenDevs. Có chức năng nhận diện giọng nói thành văn bản. Tích hợp mô hình AI Whisper của OpenAI")
                .setPositiveButton("OK", (dialog, which) -> dialog.dismiss())
                .setCancelable(true)
                .show();
    }
    private void updateLanguageSpinner() {
        String[] displayLanguages = {"Tự động", "Tiếng Việt", "English", "Japanese", "Russian", "French", "Spanish", "German", "Chinese", "Korean", "Italian"};
        String[] whisperLanguages = {"Auto", "Vietnamese", "English", "Japanese", "Russian", "French", "Spanish", "German", "Chinese", "Korean", "Italian"};
        ArrayAdapter<String> adapter = new ArrayAdapter<>(this, android.R.layout.simple_spinner_dropdown_item, displayLanguages);
        languageSpinner.setAdapter(adapter);
        int selectedIndex = java.util.Arrays.asList(whisperLanguages).indexOf(selectedLanguage);
        if (selectedIndex == -1) selectedIndex = 0;
        languageSpinner.setSelection(selectedIndex);
        languageSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                selectedLanguage = whisperLanguages[position];
                prefs.edit().putString("selectedLanguage", selectedLanguage).apply();
            }
            public void onNothingSelected(AdapterView<?> parent) {}
        });
    }
    private void showClearHistoryDialog() {
        new AlertDialog.Builder(this)
                .setTitle("Xác nhận xóa lịch sử")
                .setMessage("Bạn có chắc chắn muốn xóa toàn bộ lịch sử không? Hành động này không thể hoàn tác.")
                .setPositiveButton("Xóa", (dialog, which) -> {
                    dbHelper.clearHistory();
                    Toast.makeText(this, "Đã xóa toàn bộ lịch sử", Toast.LENGTH_SHORT).show();
                    dialog.dismiss();
                    returnToMainActivity();
                })
                .setNegativeButton("Hủy", (dialog, which) -> dialog.dismiss())
                .setCancelable(true)
                .show();
    }
    private void showResetDialog() {
        new AlertDialog.Builder(this)
                .setTitle("Xác nhận reset cài đặt")
                .setMessage("Bạn có chắc chắn muốn reset về cài đặt mặc định không?")
                .setPositiveButton("Reset", (dialog, which) -> {
                    selectedLanguage = "Vietnamese";
                    selectedModel = "Medium";
                    speechSpeed = 50;
                    speechPitch = 50;
                    serverIP.setText("14.235.42.182");

                    prefs.edit()
                            .putString("selectedLanguage", selectedLanguage)
                            .putString("selectedModel", selectedModel)
                            .putInt("speechSpeed", speechSpeed)
                            .putInt("speechPitch", speechPitch)
                            .putString("serverIP", "14.235.42.182")
                            .apply();
                    updateLanguageSpinner();
                    modelSpinner.setSelection(1);
                    speechSpeedSeekBar.setProgress(speechSpeed);
                    speechPitchSeekBar.setProgress(speechPitch);
                    Toast.makeText(this, "Đã reset về cài đặt mặc định", Toast.LENGTH_SHORT).show();
                    dialog.dismiss();
                    returnToMainActivity();
                })
                .setNegativeButton("Hủy", (dialog, which) -> dialog.dismiss())
                .setCancelable(true)
                .show();
    }
    private void returnToMainActivity() {
        Intent intent = new Intent(this, MainActivity.class);
        intent.putExtra("fromSettings", true);
        intent.putExtra("selectedLanguage", selectedLanguage);
        intent.putExtra("selectedModel", selectedModel);
        intent.putExtra("speechSpeed", speechSpeed);
        intent.putExtra("speechPitch", speechPitch);
        intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_SINGLE_TOP);
        startActivity(intent);
        finish();
    }
    @Override
    public void onBackPressed() {
        super.onBackPressed();
        returnToMainActivity();
    }
    @Override
    protected void onDestroy() {
        if (dbHelper != null) dbHelper.close();
        super.onDestroy();
    }
}

File: app/src/main/java/com/nguyendevs/stkh/settings/SettingsActivity.kt
package com.nguyendevs.stkh.settings

import android.content.Intent
import android.content.SharedPreferences
import android.os.Bundle
import android.widget.ArrayAdapter
import androidx.activity.OnBackPressedCallback
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.lifecycleScope
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.nguyendevs.stkh.MainActivity
import com.nguyendevs.stkh.R
import com.nguyendevs.stkh.database.AppDatabase
import com.nguyendevs.stkh.databinding.ActivitySettingsBinding
import com.nguyendevs.stkh.repository.IHistoryRepository
import com.nguyendevs.stkh.repository.RoomHistoryRepository
import com.nguyendevs.stkh.util.LanguageConstants
import com.nguyendevs.stkh.util.showToast
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class SettingsActivity : AppCompatActivity() {

    private lateinit var binding: ActivitySettingsBinding
    private lateinit var prefs: SharedPreferences
    private lateinit var historyRepository: IHistoryRepository

    private var selectedLanguage = "Vietnamese"
    private var selectedModel = "Medium"
    private var speechSpeed = 50
    private var speechPitch = 50

    companion object {
        private val MODELS = arrayOf("Small", "Medium", "Large")
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivitySettingsBinding.inflate(layoutInflater)
        setContentView(binding.root)

        prefs = getSharedPreferences("AppPrefs", MODE_PRIVATE)
        historyRepository = RoomHistoryRepository(AppDatabase.getInstance(this))

        restoreSettings()
        setupToolbar()
        setupLanguageSpinner()
        setupModelSpinner()
        setupSliders()
        setupServerIP()
        setupButtons()
        setupBackPressedCallback()
    }

    private fun restoreSettings() {
        selectedLanguage = prefs.getString("selectedLanguage", "Vietnamese") ?: "Vietnamese"
        selectedModel = prefs.getString("selectedModel", "Medium") ?: "Medium"
        speechSpeed = prefs.getInt("speechSpeed", 50)
        speechPitch = prefs.getInt("speechPitch", 50)
    }

    private fun setupToolbar() {
        binding.toolbar.setNavigationOnClickListener { returnToMainActivity() }
        binding.btnAbout.setOnClickListener { showAboutDialog() }
    }

    private fun setupLanguageSpinner() {
        binding.spinnerLanguage.adapter = ArrayAdapter(
            this, android.R.layout.simple_spinner_dropdown_item, LanguageConstants.DISPLAY_LANGUAGES
        )
        binding.spinnerLanguage.setSelection(LanguageConstants.WHISPER_LANGUAGES.indexOf(selectedLanguage).coerceAtLeast(0))
        binding.spinnerLanguage.onItemSelectedListener = simpleItemSelectedListener { pos ->
            selectedLanguage = LanguageConstants.WHISPER_LANGUAGES[pos]
            prefs.edit().putString("selectedLanguage", selectedLanguage).apply()
        }
    }

    private fun setupModelSpinner() {
        binding.spinnerModel.adapter = ArrayAdapter(
            this, android.R.layout.simple_spinner_dropdown_item, MODELS
        )
        binding.spinnerModel.setSelection(MODELS.indexOf(selectedModel).coerceAtLeast(0))
        binding.spinnerModel.onItemSelectedListener = simpleItemSelectedListener { pos ->
            selectedModel = MODELS[pos]
            prefs.edit().putString("selectedModel", selectedModel).apply()
        }
    }

    private fun setupSliders() {
        binding.sliderSpeechSpeed.value = speechSpeed.toFloat()
        binding.tvSpeedValue.text = speechSpeed.toString()
        binding.sliderSpeechSpeed.addOnChangeListener { _, value, fromUser ->
            if (fromUser) {
                speechSpeed = value.toInt()
                binding.tvSpeedValue.text = speechSpeed.toString()
                prefs.edit().putInt("speechSpeed", speechSpeed).apply()
            }
        }

        binding.sliderSpeechPitch.value = speechPitch.toFloat()
        binding.tvPitchValue.text = speechPitch.toString()
        binding.sliderSpeechPitch.addOnChangeListener { _, value, fromUser ->
            if (fromUser) {
                speechPitch = value.toInt()
                binding.tvPitchValue.text = speechPitch.toString()
                prefs.edit().putInt("speechPitch", speechPitch).apply()
            }
        }
    }

    private fun setupServerIP() {
        val savedIp = prefs.getString("serverIP", "14.235.42.182") ?: "14.235.42.182"
        binding.etServerIP.setText(savedIp)
        binding.etServerIP.addTextChangedListener(object : android.text.TextWatcher {
            override fun afterTextChanged(s: android.text.Editable?) {
                prefs.edit().putString("serverIP", s?.toString()?.trim()).apply()
            }
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {}
        })
    }

    private fun setupButtons() {
        binding.btnClearHistory.setOnClickListener { showClearHistoryDialog() }
        binding.btnReset.setOnClickListener { showResetDialog() }
    }

    private fun setupBackPressedCallback() {
        onBackPressedDispatcher.addCallback(this, object : OnBackPressedCallback(true) {
            override fun handleOnBackPressed() {
                returnToMainActivity()
            }
        })
    }

    private fun showAboutDialog() {
        MaterialAlertDialogBuilder(this)
            .setTitle(getString(R.string.dialog_about_title))
            .setMessage(getString(R.string.dialog_about_msg))
            .setPositiveButton(getString(R.string.action_ok)) { d, _ -> d.dismiss() }
            .show()
    }

    private fun showClearHistoryDialog() {
        MaterialAlertDialogBuilder(this)
            .setTitle(getString(R.string.dialog_clear_history_title))
            .setMessage(getString(R.string.dialog_clear_history_msg))
            .setPositiveButton(getString(R.string.action_delete)) { dialog, _ ->
                lifecycleScope.launch(Dispatchers.IO) {
                    historyRepository.clearAll()
                    runOnUiThread {
                        showToast(getString(R.string.toast_history_cleared))
                        dialog.dismiss()
                        returnToMainActivity()
                    }
                }
            }
            .setNegativeButton(getString(R.string.action_cancel), null)
            .show()
    }

    private fun showResetDialog() {
        MaterialAlertDialogBuilder(this)
            .setTitle(getString(R.string.dialog_reset_title))
            .setMessage(getString(R.string.dialog_reset_msg))
            .setPositiveButton(getString(R.string.action_reset)) { dialog, _ ->
                selectedLanguage = "Vietnamese"
                selectedModel = "Medium"
                speechSpeed = 50
                speechPitch = 50
                val defaultIp = "14.235.42.182"

                prefs.edit()
                    .putString("selectedLanguage", selectedLanguage)
                    .putString("selectedModel", selectedModel)
                    .putInt("speechSpeed", speechSpeed)
                    .putInt("speechPitch", speechPitch)
                    .putString("serverIP", defaultIp)
                    .apply()

                setupLanguageSpinner()
                setupModelSpinner()
                binding.sliderSpeechSpeed.value = speechSpeed.toFloat()
                binding.tvSpeedValue.text = speechSpeed.toString()
                binding.sliderSpeechPitch.value = speechPitch.toFloat()
                binding.tvPitchValue.text = speechPitch.toString()
                binding.etServerIP.setText(defaultIp)

                showToast(getString(R.string.toast_reset_done))
                dialog.dismiss()
                returnToMainActivity()
            }
            .setNegativeButton(getString(R.string.action_cancel), null)
            .show()
    }

    private fun returnToMainActivity() {
        val intent = Intent(this, MainActivity::class.java).apply {
            putExtra("fromSettings", true)
            putExtra("selectedLanguage", selectedLanguage)
            putExtra("selectedModel", selectedModel)
            putExtra("speechSpeed", speechSpeed)
            putExtra("speechPitch", speechPitch)
            flags = Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_SINGLE_TOP
        }
        startActivity(intent)
        finish()
    }
    private fun simpleItemSelectedListener(
        onSelected: (position: Int) -> Unit
    ): android.widget.AdapterView.OnItemSelectedListener {
        return object : android.widget.AdapterView.OnItemSelectedListener {
            override fun onItemSelected(
                parent: android.widget.AdapterView<*>?,
                view: android.view.View?,
                position: Int,
                id: Long
            ) = onSelected(position)

            override fun onNothingSelected(parent: android.widget.AdapterView<*>?) {}
        }
    }
}
